content/blog/terragrunt-envs-management.md

```diff
@@ -0,0 +1,162 @@
+
+## The Terragrunt Revelation
+
+That’s when I discovered Terragrunt. Terragrunt is a thin wrapper for Terraform that provides extra tools for keeping your Terraform code DRY (Don’t Repeat Yourself). It allows you to manage multiple Terraform modules in a single repository and reuse code across different environments. Terragrunt’s ability to manage remote state, generate configurations, and apply configurations across multiple environments made it the perfect solution for my environment management woes.
+
+How we reduce the repeated code using Terragrunt:
+First step was to make a module of the resources with all the changeable variables exposed, Next was to use them in each environment.
+
+> The new setup with Terragrunt
+
+```text
+.
+├── terragrunt.hcl
+├── production
+│   ├── app
+│   │   └── terragrunt.hcl
+│   ├── database
+│   │   └── terragrunt.hcl
+│   └── redis
+│       └── terragrunt.hcl
+├── staging
+│   ├── app
+│   │   └── terragrunt.hcl
+│   ├── database
+│   │   └── terragrunt.hcl
+│   └── redis
+│       └── terragrunt.hcl
+└── qa
+    ├── app
+    │   └── terragrunt.hcl
+    ├── database
+    │   └── terragrunt.hcl
+    └── redis
+        └── terragrunt.hcl
+```
+
+In the `terragrunt.hcl` file, we define the module to use and the
+variables to pass to the module.
+
+```hcl
+# terragrunt.hcl
+terraform {
+  source = "gittfr:///exampleorg/app/aws?version=5.8.1"
+}
+inputs = {
+  app_name = "my-app"
+  environment = "production"
+}
+```
+
+Similarly in the other terragrunt files we define the module to use and the variables to pass to the module.
+and now for the storing the state file we can use the remote state file.
+
+```hcl
+# terragrunt.hcl
+generate "backend" {
+  path      = "backend.tf"
+  if_exists = "overwrite_terragrunt"
+  contents = <<EOF
+terraform {
+  backend "s3" {
+    bucket         = "my-terraform-state"
+    key            = "${path_relative_to_include()}/terraform.tfstate"
+    region         = "us-east-1"
+    encrypt        = true
+    dynamodb_table = "my-lock-table"
+  }
+}
+EOF
+}
+```
+
+This tells terragrunt to create a new file `backend.tf` with the contents of the `contents` block, Even if backend.tf already exists it will overwrite the contents because we have specified `overwrite_terragrunt`.
+This is like a defining a function, Now we need to call it in the files i.e `terragrunt.hcl` files.
+
+> This would look like this
+
+```hcl
+include "root" {
+  path = find_in_parent_folders()
+}
+```
+
+here the `find_in_parent_folders` returns the absolute path to the first `terragrunt.hcl` file it finds in the parent folders above the current `terragrunt.hcl` file.
+and the include block tells terragrunt to include all the cofigurations from the file with the path returned by the `find_in_parent_folders` function.
+
+## The Terragrunt Workflow
+
+Here is a high-level overview of the terragrunt workflow:
+
+> We run the command `terragrunt run-all plan` in the root directory.
```

Comment:
can you add output of this command?

Reply:
Addressed in [9d07f17](https://github.com/infraspecdev/infraspec.dev/pull/58/commits/9d07f17b1bf8e7d440611789255a7ab3d6c0a140)